### DISCUSSION TITLE
Renovate automatically mark tasks as resolved in bitbucket cloud for automerge

### DISCUSSION THREAD

**Discussion by dandandy** [renovatebot/renovate#27485] [Suggest an Idea]
### Tell us more.

In bitbucket cloud you can create "tasks"(which are just a checklist https://bitbucket.org/blog/introducing-pull-request-tasks ) and PR checks which require the "tasks" to be marked as completed before merging the PR.

We are finding that renovate cannot automatically merge the PR because the tasks are not resolved (ie. ticked) which means we need to go in and manually resolve the tasks.

It would be create if renovate could be configured to automatically mark these tasks as resolved.

**Comment by rarkins** [comment]
Which tasks are you referring to? Is a developer adding them to Renovate's PRs or if not then where do they come from?

**Comment by dandandy** [comment]
I am referring to default tasks https://bitbucket.org/blog/default-pull-request-tasks (sorry I should have mentioned this important detail). So if a repository is configured with default tasks and requires the tasks to completed before merging, then renovate won't be able to merge the pull requests

**Reply by rarkins** [reply]
Thanks. I guess you are referring to this capability:

> You can add any number of default tasks to your repository configuration.
>
> Once done, these tasks will be automatically added to every new pull request.

Next questions are:
1. What happens currently? e.g. Renovate thinks it's clear to automerge, tries, gets a 4xx rejection?
2. How can Renovate learn that there's tasks to complete? e.g. can it learn from the rejection above, or better to proactively query an API before attempting automerge?
3. How can Renovate complete the tasks for its PRs? e.g. there's an API, and no special permissions necessary?

**Reply by dandandy** [reply]
Yes that's the capability I'm referring to.

So for the questions:
1. What happens currently? Unfortunately the documentation doesn't say, give me some time to test this out properly in a reproducible way.
2. How can Renovate learn that there's tasks to complete? There is a list tasks endpoint ( https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-get ). We can proactively call this endpoint to list all of the tasks. 
3. How can Renovate complete the tasks for its PRs? There is another API to update a task. It can call this endpoint to mark the tasks as resolved.

**Reply by dandandy** [reply]
Hi rankins, I have some proper answers to your questions:

1. What happens currently? What happens is that when renovate checks the status of the PR [here](https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/bitbucket/index.ts#L490) it finds the branch status is "inprogress"/"yellow", and decides that it cannot merge the PR. You can see this happen in the logs when run in debug mode (see bottom of this message on how to reproduce):
```
DEBUG: Pull Request #5 does not need updating (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: PR is configured for automerge (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: checking auto-merge (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: Checking schedule(schedule=at any time, tz=null, now=2024-07-10T07:27:54.192Z) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: No schedule defined (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: resolveBranchStatus(branchName=renovate/express-4.x, ignoreTests=false) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: getBranchStatus(renovate/express-4.x) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: branch status check result (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
       "statuses": []
DEBUG: empty branch status check result = returning "pending" (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: Branch status yellow (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: PR is not ready for merge (branch status is yellow) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
```

2. How can Renovate learn that there's tasks to complete? There [is a list tasks endpoint](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-get). 

3. How can Renovate complete the tasks for its PRs? There is [this API to update a task](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-task-id-put) and [this API to list tasks on a PR](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-get).


So I'd like to propose that we add a new renovate config called "manageBitbucketTasks" with an option for "autoClose", which will automatically closeTasks tasks whenever renovate creates/updates/observes a bitucket PR. 

---
To reproduce the automerge failure, you need to configure your bitbucket repo so that:
1. it has a "branch restriction" requiring that there are "No unresolved pull request tasks" and "Prevent a merge with unresolved merge checks" (see screenshot)
<img width="576" alt="image" src="https://github.com/renovatebot/renovate/assets/19572253/fb6996ca-15ac-4f4a-a544-a93c7041effe">

2. create a default task (see screenshot). 
<img width="1487" alt="image" src="https://github.com/renovatebot/renovate/assets/19572253/c33443a5-6feb-481c-9f60-8506d803054f">

**Reply by rarkins** [reply]
Regarding (1) the logs show that Renovate sees _no_ tests in progress or completed. It doesn't show that Renovate sees an indication of tasks. If there was one completed test (green) then I'm guessing renovate might try automerge? Is it common to have no tests?

**Reply by dandandy** [reply]
Ah, I think I misunderstood that bit of code. It was looking for a branch build status, which there was none because I hadn't set one up for that repo. But I fixed that up so now there's a green build status.

Now it looks like what happens is it tries to merge the pull request, but fails with a 400 status code, with the error message "Resolved pull request tasks":

```
DEBUG: Pull Request #5 does not need updating (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: PR is configured for automerge (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: checking auto-merge (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: Checking schedule(schedule=at any time, tz=null, now=2024-07-10T11:56:57.922Z) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: No schedule defined (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: resolveBranchStatus(branchName=renovate/express-4.x, ignoreTests=false) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: getBranchStatus(renovate/express-4.x) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: branch status check result (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
       "statuses": [
         {
           "key": "-1507272350",
           "type": "build",
           "state": "SUCCESSFUL",
           "name": "Pipeline #2 for renovate/express-4.x",
           "refname": "renovate/express-4.x",
           "commit": {
             "hash": "e5fafe5d6c7a557baae2cecab889323ec9fa30e7",
             "links": {
               "self": {
                 "href": "https://api.bitbucket.org/2.0/repositories/testing-workspace/testing-repo/commit/e5fafe5d6c7a557baae2cecab889323ec9fa30e7"
               },
               "html": {
                 "href": "https://bitbucket.org/testing-workspace/testing-repo/commits/e5fafe5d6c7a557baae2cecab889323ec9fa30e7"
               }
             },
             "type": "commit"
           },
           "url": "https://bitbucket.org/testing-workspace/testing-repo/addon/pipelines/home#!/results/2",
           "repository": {
             "type": "repository",
             "full_name": "testing-workspace/testing-repo",
             "links": {
               "self": {
                 "href": "https://api.bitbucket.org/2.0/repositories/testing-workspace/testing-repo"
               },
               "html": {"href": "https://bitbucket.org/testing-workspace/testing-repo"},
               "avatar": {
                 "href": "redacted"
               }
             },
             "name": "testing-repo",
             "uuid": "{redacted}"
           },
           "description": "",
           "created_on": "2024-07-10T11:55:45.524053+00:00",
           "updated_on": "2024-07-10T11:55:49.334213+00:00",
           "links": {
             "self": {
               "href": "https://api.bitbucket.org/2.0/repositories/testing-workspace/testing-repo/commit/e5fafe5d6c7a557baae2cecab889323ec9fa30e7/statuses/build/-1507272350"
             },
             "commit": {
               "href": "https://api.bitbucket.org/2.0/repositories/testing-workspace/testing-repo/commit/e5fafe5d6c7a557baae2cecab889323ec9fa30e7"
             }
           }
         }
       ]
DEBUG: Branch status green (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: Automerging #5 with strategy auto (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: mergePr(5, renovate/express-4.x, auto) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: POST https://api.bitbucket.org/2.0/repositories/testing-workspace/testing-repo/pullrequests/5/merge = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=400 retryCount=0, duration=1036) (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
DEBUG: PR merge error (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1720612617928,
           "socket": 1720612617929,
           "lookup": 1720612617931,
           "connect": 1720612618134,
           "secureConnect": 1720612618321,
           "upload": 1720612618322,
           "response": 1720612618959,
           "end": 1720612618964,
           "phases": {
             "wait": 1,
             "dns": 2,
             "tcp": 203,
             "tls": 187,
             "request": 1,
             "firstByte": 637,
             "download": 5,
             "total": 1036
           }
         },
         "message": "Response code 400 (Bad Request)",
         "stack": "HTTPError: Response code 400 (Bad Request)\n    at Request.<anonymous> (/usr/local/renovate/node_modules/.pnpm/got@11.8.6/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/37.426.5 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "authorization": "***********",
             "content-type": "application/json",
             "content-length": "28",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://api.bitbucket.org/2.0/repositories/testing-workspace/testing-repo/pullrequests/5/merge",
           "hostType": "bitbucket",
           "username": "",
           "password": "",
           "method": "POST",
           "http2": false
         },
         "response": {
           "statusCode": 400,
           "statusMessage": "Bad Request",
           "body": {
             "type": "error",
             "error": {
               "message": "1 failed merge check",
               "fields": {"merge_checks": ["Resolved pull request tasks"]}
             }
           },
           "headers": {
             "date": "Wed, 10 Jul 2024 11:56:59 GMT",
             "content-type": "application/json; charset=utf-8",
             "content-length": "124",
             "server": "AtlassianEdge",
             "vary": "Authorization, Origin",
             "x-accepted-oauth-scopes": "pullrequest:write",
             "x-used-mesh": "False",
             "x-credential-type": "apppassword",
             "x-token-id": "4741201",
             "x-oauth-scopes": "issue:write, pullrequest:write, team, account",
             "x-view-name": "bitbucket.apps.pullrequests.api.v20.handlers.PullRequestMergeHandler",
             "x-dc-location": "Micros-3",
             "x-served-by": "3ce3a8ebf466",
             "x-version": "1470544f5284",
             "x-static-version": "1470544f5284",
             "x-request-count": "4080",
             "x-render-time": "0.2250680923461914",
             "x-b3-traceid": "2b4de1856d4f49139c2b1d2cd3ef2d37",
             "x-b3-spanid": "bdf30871291b06b8",
             "x-frame-options": "SAMEORIGIN",
             "expires": "Wed, 10 Jul 2024 11:56:59 GMT",
             "cache-control": "max-age=0, no-cache, no-store, must-revalidate, private",
             "x-usage-user-time": "0.074151",
             "x-usage-system-time": "0.008737",
             "x-usage-input-ops": "8",
             "x-usage-output-ops": "0",
             "x-trace-id": "2b4de1856d4f49139c2b1d2cd3ef2d37",
             "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
             "api-via-stargate": "True",
             "x-content-type-options": "nosniff",
             "x-xss-protection": "1; mode=block",
             "atl-traceid": "2b4de1856d4f49139c2b1d2cd3ef2d37",
             "report-to": "{\"endpoints\": [{\"url\": \"https://dz8aopenkvv6s.cloudfront.net\"}], \"group\": \"endpoint-1\", \"include_subdomains\": true, \"max_age\": 600}",
             "nel": "{\"failure_fraction\": 0.001, \"include_subdomains\": true, \"max_age\": 600, \"report_to\": \"endpoint-1\"}",
             "connection": "close"
           },
           "httpVersion": "1.1",
           "retryCount": 0
         }
       }
 INFO: PR automerge failed (repository=testing-workspace/testing-repo, branch=renovate/express-4.x)
       "pr": 5
DEBUG: getBranchPr(renovate/express-4.x) (repository=testing-workspace/testing-repo)
```

**Reply by rarkins** [reply]
So is this the suggestion?

- Renovate can try to automerge like normal (without advanced checking)
- Renovate should look for the particular failure message and log a graceful message saying incomplete tasks prevent automerging
- If configured with [autoApprove=true](https://docs.renovatebot.com/configuration-options/#autoapprove), Renovate should then call the [list tasks endpoint](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-get) and then call the [update task endpoint](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-task-id-put) per task

**Reply by dandandy** [reply]
Yes something like that would be great

**Reply by dandandy** [reply]
Alternatively, renovate could 
1. Raise a PR in bitbucket
2. If a renovate setting is enabled (let's say it's called autoCompleteTask), then renovate will call the list task API then the update task endpoints.

This way decouples the auto merge feature from this auto task complete feature, allowing uses to control this behaviour.

Personally, for my use case, I don't mind either way, but I think what I describe in this message is more flexible and might suit other use cases better.


Let me know what you think. I'm interested in contributing code for this too

**Reply by rarkins** [reply]
I agree with your idea, PR welcome

**Reply by viceice** [reply]
we've a similar auto complete flag for azure

**Reply by dandandy** [reply]
I have raised a PR here https://github.com/renovatebot/renovate/pull/30901
